Feature/professor lol 24

csbsjy opened the pull request:
형식 자체는 민정이 틀 따라가려고 노력해서 이해안가는 부분은 없을거얌 .. 
커밋메시지는 미안 실수 ㅎ;;

csbsjy committed:
1. stram().forEach() -> forEach() 수정
2. 테스트코드 assertThat 추가

csbsjy committed:
1. stram().forEach() -> forEach() 수정
2. 테스트코드 assertThat 추가

csbsjy committed:
test

csbsjy committed:
test

Gobukgol commented on professorlol-crawling/src/main/java/com/css/professorlol/dto/champion/NewChampionFactory.java:
```diff
@@ -0,0 +1,29 @@
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
+public class NewChampionFactory {
+
+    private static final String CONTENT_BORDER = ".content-border .patch-change-block";
```
저렇게 하면 .content-border 랑 .patch-chage-block 으로 되어있는거 각각 가져오는거야?

jyami-kim replied:
ㅇㅇ

chldbtjd2272 commented on professorlol-crawling/src/main/java/com/css/professorlol/RiotPageBugFixCrawler.java:
```diff
@@ -0,0 +1,14 @@
+public class RiotPageBugFixCrawler implements BugFixPatchCrawler {
+    @Override
+    public List<BugFix> findBugFixPatchByDocument(Document document) {
+        return new BugFixFactory(document).getBugFixes();
```
Factory를 멤버변수로 들고있어야해 
매번 부를때마다 Factory를 생성하면 안좋아

csbsjy replied:
이렇게 수정했어! 
```java
public class RiotPageBugFixCrawler implements BugFixPatchCrawler {

    private BugFixParser bugFixParser;

    public RiotPageBugFixCrawler(BugFixParser bugFixParser) {
        this.bugFixParser = bugFixParser;
    }

    @Override
    public List<BugFix> findBugFixPatchByDocument(Document document) {
        return bugFixParser.getBugFixes();
```
    }
}

chldbtjd2272 commented on professorlol-crawling/src/main/java/com/css/professorlol/dto/bug/BugFixFactory.java:
```diff
@@ -0,0 +1,25 @@
+
+    public BugFixFactory(Document document) {
+        Elements elements = document.select(CONTENT_BORDER);
+        this.bugFixes = elements.stream()
```
bugFixes가 멤버변수일 필요는 없을듯
Elements객체가 멤버변수로 있는게 맞을거같아

csbsjy replied:
아래처럼 수정했고, 네이밍도 Parser로 전부 변경했어.
파싱하는 로직은 메서드로 이동했어!
```java
    private final Elements elements;

    public BugFixParser(Document document) {
        elements = document.select(CONTENT_BORDER);
    }
```

chldbtjd2272 commented on professorlol-crawling/src/main/java/com/css/professorlol/dto/champion/NewChampionMapper.java:
```diff
@@ -0,0 +1,35 @@
+import static com.css.professorlol.dto.champion.NewChampionMapper.NewChampionSelector.*;
+import static com.css.professorlol.util.ElementContextUtil.*;
+
+public class NewChampionMapper {
```
인터페이스로 빼고 enum을 인터페이스에 정의

csbsjy replied:
고민해 봤는데, convert 메서드의 리턴값인 NewChampion과 Champion Dto가 완전히 다른 객체여서 인터페이스로 분리했을 때 가질 수 있는게 enum밖에 없더라고! 인터페이스로 분리하기보다는 그냥 enum클래스를 밖으로 빼는게 더 괜찮아 보여서 일단 그렇게 했어!

chldbtjd2272 commented on professorlol-crawling/src/main/java/com/css/professorlol/util/ElementContextUtil.java:
```diff
@@ -32,5 +37,14 @@ public static String getUrlFromCssSelector(Elements elements, String cssSelector
         return getUrlFromElement(elements.select(cssSelector));
     }
 
+    public static List<LinkGroup> getUrlListFromCssSelector(Element element, String cssSelector) {
```
util클래스가 dto를 생성하면 영향도가 너무커

csbsjy replied:
해당 로직은 mapper로 이동했고, 유틸클래스에 링크가 존재하는지 체크 메서드 하나 추가했어
```java
    public static String getLinkElement(Element element){
        Elements liElements = element.select(LINK_ATTRIBUTE_KEY);
        return liElements.isEmpty() ? "" : liElements.select(LINK_ATTRIBUTE_KEY).attr(URL_VALUE_ATTRIBUTE_KEY);
    }
```

chldbtjd2272 commented on professorlol-crawling/src/main/java/com/css/professorlol/util/ElementContextUtil.java:
```diff
@@ -32,5 +37,14 @@ public static String getUrlFromCssSelector(Elements elements, String cssSelector
 
+    public static List<LinkGroup> getUrlListFromCssSelector(Element element, String cssSelector) {
+        List<LinkGroup> urlList = new ArrayList<>();
+        element.select(cssSelector).forEach(liElement -> urlList.add(
```
map으로 바로 반환하면 좋을듯

csbsjy replied:
Mapper클래스로 이동하고 아래처럼 다시 구현했어! 링크 리스트가 없는 경우 빈 ArrayList를 반환하도록 했는데 괜찮은 건지 모르겠당
```java
private static List<LinkGroup> getUrlListFromCssSelector(Element element, String cssSelector) {
        return element.select(cssSelector).isEmpty() ? new ArrayList<>() : element.select(cssSelector).stream()
                .map(liElement -> LinkGroup.builder()
                        .content(liElement.text())
                        .url(getLinkElement(liElement))
                        .build())
                .collect(Collectors.toList());
    }
```

pci2676 replied:
우리 코드 컨벤션이 depth1번까지 허용이니까 
map 내부 람다식에서 depth가 2번 들어가니까 1번으로 들어갈 수 있게 함수로 빼는건어때? ㅎㅎ

chldbtjd2272 commented on professorlol-crawling/src/main/java/com/css/professorlol/util/ElementContextUtil.java:
```diff
@@ -32,5 +37,14 @@ public static String getUrlFromCssSelector(Elements elements, String cssSelector
+        element.select(cssSelector).forEach(liElement -> urlList.add(
+                LinkGroup.builder()
+                        .content(liElement.text())
+                        .url(liElement.select("a").attr(URL_LINK_ATTRIBUTE_KEY))
```
a태그 체크하는건 util에

chldbtjd2272 commented on professorlol-crawling/src/test/java/PatchNoteCrawlerMockTest.java:
```diff
@@ -44,6 +44,9 @@ public void tearDown() {
         PatchNoteCrawler patchNoteCrawler = new RiotPagePatchNoteCrawler(createMockJsoupConnection());
         List<Champion> champions = patchNoteCrawler.findChampionPatchById(id);
 
+        champions.stream()
```
제거

chldbtjd2272 commented on professorlol-crawling/src/test/java/com/css/professorlol/dto/bug/BugFixFactoryTest.java:
```diff
@@ -0,0 +1,35 @@
+    public void getBugFixes() {
+        BugFixFactory bugFixFactory = new BugFixFactory(document);
+        List<BugFix> bugFixes = bugFixFactory.getBugFixes();
+        assertEquals(bugFixes.get(0).getContext(),"나르가 미니언에게 도약하기 전 R - 나르!로 변신한 후 E - 폴짝의 방향을 바꿀 수 있었던 버그를 수정했습니다.");
```
assertThat().isEqualTo()이거 사용
junit core쪽에 있는 함수

csbsjy replied:
이유가뭐야?

pci2676 replied:
그게 더 읽기가 편해 core쪽에 있는 애들이 더 명시적으로 표현이 되기도하구

chldbtjd2272 commented on professorlol-crawling/src/test/java/RiotNewChampionPatchCrawlerTest.java:
```diff
@@ -0,0 +1,29 @@
+    private static final String PATCH_NOTE_0916_HTML = "PatchNote_9_16.html";
+
+    @Test
+    public void Document_객체에서_크롤링되는것_확인() {
```
예외 상황에 대한 테스트가없어

csbsjy replied:
신챔이나 버그가 없는 경우에 빈 리스트가 반환되는 테스트를 하나 추가하긴 했는데 더 생각해볼ㄱㅔ~~

chldbtjd2272 commented on professorlol-crawling/src/main/java/com/css/professorlol/RiotNewChampionPatchCrawler.java:
```diff
@@ -0,0 +1,16 @@
+
+    @Override
+    public List<NewChampion> findNewChampionPatchByDocument(Document document) {
+        return new NewChampionFactory(document).getNewChampions();
```
펙토리를 멤버변수로

pci2676 commented on professorlol-crawling/src/main/java/com/css/professorlol/dto/bug/BugFixFactory.java:
```diff
@@ -0,0 +1,25 @@
+    public BugFixFactory(Document document) {
+        Elements elements = document.select(CONTENT_BORDER);
+        this.bugFixes = elements.stream()
+                .filter(element -> element.select("a").size() == 0) //신챔 제외
```
이게 신챔은 제외한다는 걸 주석말고 함수로 알수 있게 빼는건 어떻게 생각해?

pci2676 commented on professorlol-crawling/src/main/java/com/css/professorlol/dto/champion/NewChampionFactory.java:
```diff
@@ -0,0 +1,29 @@
+        Elements contentBorder = document.select(CONTENT_BORDER);
+
+        this.newChampions = contentBorder.stream()
+                .filter(element -> element.select("li a").size() > 0) //신챔만 골라내기
```
여기두 신챔을 골라낸다는걸 코드만 보고는 이해하지 못할거 같아

pci2676 commented on professorlol-crawling/src/main/java/com/css/professorlol/dto/champion/NewChampionMapper.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    //TODO: ChampionSelector랑 공통되는 부분이 많은데, Enum 을 따로 빼서 관리하는게 낫지 않을지 고민임!
+    protected enum NewChampionSelector {
```
TODO에 적어둔 것처럼 Selector가 겹치는게 많으니까 공통된 enum으로 빼도 될거같아.
근데이게 우연한 중복인지 진짜 중복인지는 좀더 생각해 봐야할것같아

pci2676 commented:
리뷰 반영된것들 푸시해줘~

csbsjy committed:
리뷰 반영

pci2676 commented on professorlol-crawling/src/main/java/com/css/professorlol/dto/champion/ChampionMapper.java:
```diff
@@ -1,37 +1,22 @@
                 .context(getTextFromCssSelector(element, CONTEXT.cssQuery))
                 .image(getUrlFromCssSelector(element, IMAGE.cssQuery))
-                .abilities(AbilityFactory.of(element))
+                .abilities(AbilityParser.of(element))
```
보다가 느낀건데 of() 하면 element에서 AbilityParser 객체가 생성 될것같이 느껴져
parse()같은거는 어때?

Gobukgol commented on professorlol-crawling/src/main/java/com/css/professorlol/RiotPageBugFixCrawler.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Override
+    public List<BugFix> findBugFixPatchByDocument(Document document) {
```
이 함수에서 document는 어디에 쓰이는거야??? 인자로 받는데 사용하지 않는걸로 보여서!

jyami-kim commented on professorlol-crawling/src/main/java/com/css/professorlol/dto/bug/BugFixParser.java:
```diff
@@ -0,0 +1,26 @@
+
+    public List<BugFix> getBugFixes() {
+        return elements.stream()
+                .filter(element -> element.select("a").size() == 0) //신챔 제외
```
"a" 태그 가리키는거 상수로 빼면 이상한가?!
뒤에는 상수로 뺐길랭!

jyami-kim commented on professorlol-crawling/src/main/java/com/css/professorlol/dto/champion/NewChampionParser.java:
```diff
@@ -0,0 +1,25 @@
+
+    public List<NewChampion> getNewChampions() {
+        return contentBorder.stream()
+                        .filter(element -> element.select("li a").size() > 0) //신챔만 골라내기
```
상수로!

jyami-kim commented on professorlol-crawling/src/main/java/com/css/professorlol/util/ElementContextUtil.java:
```diff
@@ -32,5 +34,9 @@ public static String getUrlFromCssSelector(Elements elements, String cssSelector
 
+    public static String getLinkElement(Element element){
+        Elements liElements = element.select(LINK_ATTRIBUTE_KEY);
+        return liElements.isEmpty() ? "" : liElements.select(LINK_ATTRIBUTE_KEY).attr(URL_VALUE_ATTRIBUTE_KEY);
```
위에 상수로 뺀 EMPTY_STRING 으로 바꾸면 좋을 것 같아

jyami-kim commented on professorlol-crawling/src/test/java/RiotNewChampionPatchCrawlerTest.java:
```diff
@@ -0,0 +1,42 @@
+        //given
+        File resourceFile = getFileFromPath(PATCH_NOTE_0916_HTML_EXCLUDE_NEW_CHAMPION);
+        Document document = convertFromHtmlFile(resourceFile);
+        List<NewChampion> newChampions = new RiotNewChampionPatchCrawler(new NewChampionParser(document)).findNewChampionPatchByDocument(document);
```
new ( new (...)) 이런식으로 괄호에 괄호 이렇게 있으니까 읽기 조금 어려운듯!!

jyami-kim commented on professorlol-crawling/src/test/java/RiotPageBugFixCrawlerTest.java:
```diff
@@ -0,0 +1,43 @@
+    public void 버그패치가_없는_경우_빈_리스트_출력되는것_확인(){
+        File resourceFile = getFileFromPath(PATCH_NOTE_0916_HTML_EXCLUDE_BUG);
+        Document document = convertFromHtmlFile(resourceFile);
+        List<BugFix> bugFixes = new RiotPageBugFixCrawler(new BugFixParser(document)).findBugFixPatchByDocument(document);
```
위에랑 같은이유로 new(new(..)) 겹쳐있어서 조금 어렵쓰!
옛날에 내가 그렇게 썼었나ㅋㅋㅋ

csbsjy committed:
리뷰반영

temp

professor-lol-24

1. 크롤러 변경사항 없으면 new로 들어가도록 수정

professor-lol-24

1. 크롤러 변경사항 없으면 new로 들어가도록 수정

jyami-kim commented:
Factory > Parser
BugFixParser: constructor에 document들어가있는거

Parser에 of() > parse()

Mapper Parser 모두 static method끼리 물고있어서 어떻게 돌아가는지 모르겠다.
Parser 안에 Mapper그냥 넣어두기
